client/alloc_runner_health_watcher.go

```diff
@@ -31,31 +31,29 @@ func (r *AllocRunner) watchHealth(ctx context.Context) {
 
 	// See if we should watch the allocs health
 	alloc := r.Alloc()
-	if alloc.DeploymentStatus.IsHealthy() || alloc.DeploymentStatus.IsUnhealthy() {
-		// No need to watch health as it's already set
-		return
-	}
 
-	// Neither deployments nor migrations care about system jobs so never
-	// watch their health
-	if alloc.Job.Type == structs.JobTypeSystem {
+	// Neither deployments nor migrations care about the health of
+	// non-service jobs so never watch their health
+	if alloc.Job.Type != structs.JobTypeService {
 		return
 	}
 
-	isDeploy := alloc.DeploymentID != ""
-
-	// Migrations don't consider the health of batch jobs so only watch
-	// batch health during deployments
-	if !isDeploy && alloc.Job.Type == structs.JobTypeBatch {
+	// No need to watch health as it's already set
+	if alloc.DeploymentStatus.IsHealthy() || alloc.DeploymentStatus.IsUnhealthy() {
 		return
 	}
 
 	tg := alloc.Job.LookupTaskGroup(alloc.TaskGroup)
 	if tg == nil {
-		r.logger.Printf("[ERR] client.alloc_watcher: failed to lookup allocation's task group. Exiting watcher")
+		r.logger.Printf("[ERR] client.alloc_watcher: failed to lookup allocation %q task group %q. Exiting watcher",
+			alloc.ID, alloc.TaskGroup)
 		return
 	}
 
+	isDeploy := alloc.DeploymentID != ""
+
+	// No need to watch health during updates without deployments OR
+	// deployments that rely on manually setting health
 	if isDeploy && (tg.Update == nil || tg.Update.HealthCheck == structs.UpdateStrategyHealthCheck_Manual) {
```

Comment:
Splitting this into two case statements might make the logic cleaner, as it seems appear fairly unrelated?

Reply:
Hm, just realized the first half of my comment is nonsense:

> No need to watch health during updates without deployments

The following `if` statement starts with `if isDeploy`. The whole clause only deals with deployments. I think the `tg.Update==nil` check just confused me. I thought it was a variation on deploys when really it's just to guard prevent the `tg.Update.HealthCheck` from panic'ing.

I'll fixup the comment as it only handles the second clause.